File: 206/python/solutions.py
from lib.lists.linkedlists import SinglyLinkedList
from nodes import ListNode


class Solution1:
    def reverseList(self, head: ListNode) -> ListNode:
        return SinglyLinkedList(head).reverse()


class Solution2:
    # Recursion
    def reverseList(self, head: ListNode) -> ListNode:
        return self._helper(head)

    def _helper(self, head: ListNode, prev: ListNode = None) -> ListNode:
        if not head: return prev
        next = head.next;
        head.next = prev
        return self._helper(next, head)

File: 206/cpp/solutions.cpp
#include <stddef.h>

#include "lib/lists/linkedlists.h"
#include "nodes.h"

using namespace dsa::lib;
using namespace dsa::nodes;

class Solution1
{
public:
    ListNode *reverseList(ListNode *head)
    {
        SinglyLinkedList *ls;
        ls = new SinglyLinkedList(head);
        return ls->reverse();
    }
};

class Solution2
{
    // Recursion
public:
    ListNode *reverseList(ListNode *head)
    {
        if (!head || !(head->next))
            return head;
        ListNode *newHead = reverseList(head->next);
        ListNode *tail = head->next;
        tail->next = head;
        head->next = NULL;
        return newHead;
    }
};

class Solution3
{
public:
    ListNode *reverseList(ListNode *head)
    {
        return recursion(head, NULL);
    }

    ListNode *recursion(ListNode *head, ListNode *prev)
    {
        if (!head)
            return prev;
        ListNode *next = head->next;
        head->next = prev;
        return recursion(next, head);
    }
};